Sources/MongoSwift/BSON/BSON.swift

```diff
@@ -0,0 +1,413 @@
+import Foundation
+
+/// Enum representing a BSON value.
+/// - SeeAlso: bsonspec.org
+public enum BSON {
+    /// A BSON double.
+    case double(Double)
+
+    /// A BSON string.
+    /// - SeeAlso: https://docs.mongodb.com/manual/reference/bson-types/#string
+    case string(String)
+
+    /// A BSON document.
+    case document(Document)
+
+    /// A BSON array.
+    indirect case array([BSON])
+
+    /// A BSON binary.
+    case binary(Binary)
+
+    /// A BSON undefined.
+    case undefined
+
+    /// A BSON ObjectId.
+    /// - SeeAlso: https://docs.mongodb.com/manual/reference/bson-types/#objectid
+    case objectId(ObjectId)
+
+    /// A BSON boolean.
+    case bool(Bool)
```

Comment:
I'm noticing we have some naming mismatches between this enum and the `BSONType` enum:
- `bool` vs `boolean`
- `date` vs `dateTime`
- `regex` vs `regularExpression`
- `javascript` vs `code`
- `javascriptWithScope` vs `codeWithScope`

I don't feel very strongly about either option in any of these cases but it does seem to me like they should be consistent.

Reply:
Good point. I've decided on the following:
- `.bool`: expresses the same idea as "boolean" but is slightly briefer
- `.datetime`: since dates and datetimes are different things, I felt it was important to be precise here. I did away with the camel case since a datetime is thing itself rather than a combination of two words.
- `.regex`: communicates the same idea as "regularExpression" but is much briefer
- `.code` and `.codeWithScope`: "code" is much briefer than "javascript" but still stays true to the spec

Let me know if this scheme makes sense. I also updated some of the docstrings to be more descriptive.